DOCS: state that `initial_properties` are shared between entity instances (#14945)
minetest/minetest: doc/lua_api.md

Entity definition
-----------------

Used by `minetest.register_entity`.  
The entity definition table becomes a metatable of a newly created per-entity

> luaentity table, meaning its fields (e.g. `initial_properties`) will be shared

Reviewer: I understand this has been merged, but the wording feels a bit confusing to me. When it says shared between all instances, this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.

As I understood it, the initial properties are used to set up a new instance of the entity (pass by value).

Wording such as: "The entity definition table is used in the creation of a metatable for a per-entity luaentity table, meaning its fields (e.g. `initial_properties`) will become the default values for each newly created instance of the entity." would be more clear.

If I have misunderstood, please let me know as I'm curious now.

Reviewer: I just stumbled across the discussion after this, and felt Small Joker's wording was less ambigous:

"The entity definition table becomes a metatable of a newly created per-entity
luaentity table, roughly meaning its fields (e.g. `initial_properties`) will be
provided as default values."

Reviewer: I think some of the discussion implies that modders should know Lua deeply, and I do not think that is the case for most people modding on minetest. They come in to it from different languages and different background. metatables aren't something known in many languages, but by reference and value is.

They want to make games and this could be the first place they learn lua.

Reviewer 2: > this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.

this is correct.
what might be confusing you is that `initial_properties` is read by the engine once on object creation. any further property changes have nothing to do with this object, there isn't even a property table stored in Lua.
but if you edit the `initial_properties` table it will indeed affect all future instances.

Reviewer: > > this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.
> 
> this is correct. what might be confusing you is that `initial_properties` is read by the engine once on object creation. any further property changes have nothing to do with this object, there isn't even a property table stored in Lua. but if you edit the `initial_properties` table it will indeed affect all future instances.

Interesting. Appreciate you taking the time to clarify this.

So when you set properties, this will have some form of override mechanism, or simply set up on the object (which was set on creation)?

Reviewer 3: > When it says shared between all instances, this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.

This is the case. If you do `entity1.initial_properties.pointable = false` (which is not something you should do), then `entity2.initial_properties.pointable` will suddenly also be `false`, even if it was true before (if `entity1` and `entity2` are of the same type). This is because `entity1.initial_properties == entity2.initial_properties` which is just `minetest.registered_entities[entity1.name].initial_properties`.

> As I understood it, the initial properties are used to set up a new instance of the entity (pass by value).

Yes, on the C++ side of things, but that affects `self.object:[gs]et_properties`, not `self.initial_properties`.

---

Honestly at this point I think it'd probably be best to just specify exactly what we do, along the lines of:

> When Minetest instantiates an entity `self`, it sets the entity definition table `def` as the metatable ("prototype") of the object. The entity definition table has its `__index` field referring to itself. (This means that if the `self` table does not directly contain a key `k`, `self[k]` will evaluate to `def[k]`. For details refer to the Lua reference manual on metatables.)

But generally I don't think we should be explaining metatables *here* (if anything, the concept of prototypical OOP / metatables should be introduced as a term in a sort of "Lua primer" along with other terms which come up, and then be referenced here).

---

> I think some of the discussion implies that modders should know Lua deeply

Metatables are pretty fundamental, and existing resources like the PIL and the reference manual (and probably many other) already explain them well enough. They aren't that hard to understand. And similar prototypical OOP is found in e.g. JavaScript.

> but by reference and value is

Yes, but unfortunately not exactly applicable here. There is no way to be precise about what's going on here without speaking of metatables (at least the `__index` metamethod / field). This requires that the user knows what a metatable with `__index` set does.

Reviewer: > > When it says shared between all instances, this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.
> 
> This is the case. If you do `entity1.initial_properties.pointable = false` (which is not something you should do), then `entity2.initial_properties.pointable` will suddenly also be `false`, even if it was true before (if `entity1` and `entity2` are of the same type). This is because `entity1.initial_properties == entity2.initial_properties` which is just `minetest.registered_entities[entity1.name].initial_properties`.
> 
> > As I understood it, the initial properties are used to set up a new instance of the entity (pass by value).
> 
> Yes, on the C++ side of things, but that affects `self.object:[gs]et_properties`, not `self.initial_properties`.
> 
> Honestly at this point I think it'd probably be best to just specify exactly what we do, along the lines of:
> 

Thanks for clarifying

> > When Minetest instantiates an entity `self`, it sets the entity definition table `def` as the metatable ("prototype") of the object. The entity definition table has its `__index` field referring to itself. (This means that if the `self` table does not directly contain a key `k`, `self[k]` will evaluate to `def[k]`. For details refer to the Lua reference manual on metatables.)
> 

That's pretty clear imho.

> But generally I don't think we should be explaining metatables _here_ (if anything, the concept of prototypical OOP / metatables should be introduced as a term in a sort of "Lua primer" along with other terms which come up, and then be referenced here).
> 

This is a really good idea.

> > I think some of the discussion implies that modders should know Lua deeply
> 
> Metatables are pretty fundamental, and existing resources like the PIL and the reference manual (and probably many other) already explain them well enough. They aren't that hard to understand. And similar prototypical OOP is found in e.g. JavaScript.
> 

I'll probably shame myself in saying this, but I do not understand metatables well. I do think I would like to do some more reading. When I started reading about them, it did feel a bit confusing. I'm really confident in OOP in say Java, and languages like it, but when it comes to metatables, I think it unclear and confusing. Maybe I need to revisit it.

I've been a developer using OOP for over a decade and it isn't as easy as you imply. I think the reality is you are ridiculously intelligent and with the CS background, it makes things seem easy that probably are not. I really enjoy when you go into detail and explain stuff as I always learn things. I'm in a group of modders that uses a language to achieve things, and I'm usually successful in it. I enjoy learning more, but if learning more doesn't seem to have a tangible benefit to my ability to make things, I struggle with motivation. Long covid brain doesn't help that also, I guess.

I always find javascript hard to read. The syntax doesn't seem to work well with my brain. I can change it, but I'm not expressive in it, and it does take some effort.

> > but by reference and value is
> 
> Yes, but unfortunately not exactly applicable here. There is no way to be precise about what's going on here without speaking of metatables (at least the `__index` metamethod / field). This requires that the user knows what a metatable with `__index` set does.

I appreciate the clarification, and probably do need to dig into it further.

Thanks for your time and the response on this.

Reviewer 3: > This is a really good idea.

I have opened a draft PR: https://github.com/minetest/minetest/pull/15112

> The entity definition table becomes a metatable of a newly created per-entity
> table, meaning its subtables (e.g. `initial_properties`) will be shared between
> all instances of an entity

Reviewer 4: ```suggestion
The entity definition table becomes a metatable of a newly created per-entity
luaentity table, meaning its fields (e.g. `initial_properties`) will be shared
between all instances of an entity.
```
The fields are not just tables.

Also, mentioning subtables makes it sound to me like you had metatable semantics *in* the subtables, which is not the case. If you override a field in `luaentity.initial_properties`, you modify the prototype.

Btw., FYI, but probably not worth mentioning, being set as a metatable isn't enough make the index magic work. What also happens is that the `__index` field is set (see `prototype.__index = prototype` in `core.register_entity` in builtin).

Author: With your suggestion it sounds like every field is shared, which is not the case. If I have `foo = 5` as a custom field, that won't be shared. If I have `bar = { foo = 5 }` that will

Reviewer 5: Relevant code: 

https://github.com/minetest/minetest/blob/1222750c50c3077d5793675bfdc337b66ed24313/src/script/cpp_api/s_entity.cpp#L44-L70

`core.registered_entities[name]` is used as a metatable of the new entity table. Perhaps it's the most technically correct way to describe it as

```suggestion
The entity definition table becomes a metatable of a newly created per-entity
luaentity table, roughly meaning its fields (e.g. `initial_properties`) will be
provided as default values.
```

Reviewer 2: It is correct but can be confusing to people less familar with Lua. 

in the following example you won't have two copies of the string:
```lua
local a = "Minetest"
local b = a
```
it's just that strings are immutable so there's no way to mess it up

Reviewer 4: >(or rather: everything except tables) are immutable

Functions (via `setfenv`), userdata, and coroutine threads are also mutable, so it's not just tables.

But I can see how my suggestion can cause confusion.

Krock's suggestion sounds good to me.

Reviewer 2: > Krock's suggestion sounds good to me.

It emphasizes the "sharing" aspect even less, however.

Reviewer 2: So do we want to change this? I like the initial suggestion from @Desour.

Author: I don't know, I personally find it confusing, but if two core devs agree who am I to say otherwise

Reviewer 4: You can also make it longer and more explicit, e.g.:
```suggestion
The entity definition table becomes a metatable of a newly created per-entity
luaentity table, roughly meaning its fields (e.g. `initial_properties`) will be
provided as default values (but *not* deep copied).
```

If the modder knows lua well enough, they'll probably understand either way, because they know how metatables and values work.

Author: Went for the first one, simpler

    -- A table of object properties, see the `Object properties` section.
    -- The properties in this table are applied to the object
    -- once when it is spawned.

> 	-- Contrary to custom fields down below, initial_properties is a table
> 	-- shared between all the instances of an entity. Changing any of its
> 	-- fields on one instance will alter any other present and future instance
> 	-- as well

Reviewer 2: I would get right to the point and reword this like so:

```suggestion
	-- Do not try to change this table at runtime, because it is shared
	-- between all instances of an entity.
```

Author: I think it's important to underline that this behaves differently from custom fields, which is what it took me some time to realise

Reviewer 2: I would solve that by adding a paragraph at the top explaining how the entity definition table becomes a metatable of a newly created per-entity table.

Author: Wait, so if I have a custom field that is a table, is that shared as well?

Author: Should be fine now
